modules/020-deckhouse/hooks/generate_selfsigned_ca.go

```diff
@@ -20,97 +20,38 @@ import (
 	"fmt"
 
 	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
-	"github.com/flant/addon-operator/sdk"
-	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
-	v1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 
-	"github.com/deckhouse/deckhouse/go_lib/certificate"
+	hooks "github.com/deckhouse/deckhouse/go_lib/hooks/internal_tls"
 )
 
-func applyCertFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
-	secret := &v1.Secret{}
-	err := sdk.FromUnstructured(obj, secret)
-	if err != nil {
-		return nil, fmt.Errorf("cannot convert selfsigned certificate secret to secret: %v", err)
-	}
-
-	return certificate.Certificate{
-		CA:   string(secret.Data["ca.crt"]),
-		Key:  string(secret.Data["tls.key"]),
-		Cert: string(secret.Data["tls.crt"]),
-	}, nil
-}
-
-var _ = sdk.RegisterFunc(&go_hook.HookConfig{
-	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
-	Kubernetes: []go_hook.KubernetesConfig{
-		{
-			Name:       "cert",
-			ApiVersion: "v1",
-			Kind:       "Secret",
-			NamespaceSelector: &types.NamespaceSelector{
-				NameSelector: &types.NameSelector{
-					MatchNames: []string{"d8-system"},
-				},
-			},
-			NameSelector: &types.NameSelector{
-				MatchNames: []string{"webhook-handler-certs"},
-			},
-			FilterFunc: applyCertFilter,
-		},
-	},
-}, generateSelfSignedCertificate)
-
-func generateSelfSignedCertificate(input *go_hook.HookInput) error {
-	const (
-		webhookServiceHost = "webhook-handler.d8-system.svc"
-
-		webhookHandlerCertPath = "deckhouse.internal.webhookHandlerCert.crt"
-		webhookHandlerKeyPath  = "deckhouse.internal.webhookHandlerCert.key"
-		webhookHandlerCAPath   = "deckhouse.internal.webhookHandlerCert.ca"
-	)
-
-	var sefSignedCert certificate.Certificate
-
-	certs := input.Snapshots["cert"]
-	if len(certs) == 1 {
-		var ok bool
-		sefSignedCert, ok = certs[0].(certificate.Certificate)
-		if !ok {
-			return fmt.Errorf("cannot convert sefsigned certificate to certificate")
-		}
-	} else {
-		var err error
-		sefSignedCA, err := certificate.GenerateCA(input.LogEntry, webhookServiceHost)
-		if err != nil {
-			return fmt.Errorf("cannot generate selfsigned ca: %v", err)
-		}
+const (
+	webhookServiceHost = "webhook-handler.d8-system.svc"
+)
 
+var _ = hooks.RegisterInternalTLSHook(hooks.GenSelfSignedTLSHookConf{
```

Comment:
Looks good! But i should remind that old hook generated certificate for 1 year, `go_lib/hooks/internal_tls/hook.go` generates certificate for 10 years. I think it is not a problem but it changes hook semantic.

Reply:
@nabokihms I do not know, is it problem for us?

Reply:
It was a main goal of this patch :) self-signed certificate for webhook handler was created for 1 year when it should be created for 10 years.
All other changes - it's a try to make generic solution for working with self-signed certificates, with an expiration check also

Reply:
oh, ok. i've resolved this